src/generic/gentpcc.tcl

```diff
@@ -4,26 +4,22 @@ allwarehouse {
 #set additional text for all warehouses
 set allwt(1) {set allwarehouses "true";# Use all warehouses to increase I/O
 }
-set allwt(2) {#2.4.1.1 does not apply when allwarehouses is true 
+set allwt(2) {#2.4.1.1 does not apply when allwarehouses is true
 if { $allwarehouses == "true" } {
-set mypos [ expr $myposition -1 ]
-if { $mypos > $w_id_input  } { 
-set mymod [expr ($mypos % $w_id_input) ]
-if { $mymod eq 0 }  { set mymod $w_id_input } 
-lappend myWarehouses $mymod
-} else {
 set loadUserCount [expr $totalvirtualusers - 1]
 set myWarehouses {}
-lappend myWarehouses [expr $myposition - 1]
-set addMore 1
-while {$addMore > 0} {
-set wh [expr ($myposition - 1) + ($addMore * $loadUserCount)]
-if {$wh > $w_id_input || $wh eq 1} {
 set addMore 0
+set whRequiredCount [expr ceil(double($w_id_input)/double($loadUserCount))]
+while {$addMore >= 0} {
+set wh [expr ((($myposition - 2) + ($addMore * $loadUserCount))%$w_id_input)+1]
+if {$addMore >= $whRequiredCount} {
+set addMore -1
 } else {
+puts "VU $myposition : Assigning WID=$wh based on VU count $loadUserCount, Warehouses = $w_id_input ($addMore out of $whRequiredCount)"
```

Comment:
This all looks good. In testing the orginal the assigned warehouses can miss the first warehouse or group of warehouses and this resolves the issue. 

Only suggestion is to modify the output line to the following: 

`puts "VU $myposition : Assigning WID=$wh based on VU count $loadUserCount, Warehouses = $w_id_input ([expr $addMore + 1] out of [ expr int($whRequiredCount)])"`

At the moment this changes prints the following: 

Vuser 6:VU 6 : Assigning WID=5 based on VU count 5, Warehouses = 20 (0 out of 4.0)
Vuser 6:VU 6 : Assigning WID=10 based on VU count 5, Warehouses = 20 (1 out of 4.0)
Vuser 6:VU 6 : Assigning WID=15 based on VU count 5, Warehouses = 20 (2 out of 4.0)
Vuser 6:VU 6 : Assigning WID=20 based on VU count 5, Warehouses = 20 (3 out of 4.0)

and this changes it to the more user friendly. 

Vuser 6:VU 6 : Assigning WID=5 based on VU count 5, Warehouses = 20 (1 out of 4)
Vuser 6:VU 6 : Assigning WID=10 based on VU count 5, Warehouses = 20 (2 out of 4)
Vuser 6:VU 6 : Assigning WID=15 based on VU count 5, Warehouses = 20 (3 out of 4)
Vuser 6:VU 6 : Assigning WID=20 based on VU count 5, Warehouses = 20 (4 out of 4)

if you can make this change I will merge the pull request